billing-db/sprocs/get_tenant_bill.sql

```diff
@@ -189,3 +193,278 @@ BEGIN
+    -- Case 0:
+    -- _from_date, _to_date:                  |---------------------------|
+    -- Resource present:            |-----------------|
+    SELECT  GREATEST(_from_date, r.valid_from),
+            LEAST(_to_date, r.valid_to),
+            r.resource_guid,
+            r.resource_type,
+            r.resource_name,
+            r.org_guid,
+            r.org_name,
+            r.space_guid,
+            r.space_name,
+            r.plan_name,
+            r.plan_guid,
+            r.storage_in_mb,
+            r.memory_in_mb,
+            r.number_of_nodes
+    FROM  resources r
+    WHERE r.valid_from < _from_date
+    AND   r.valid_to > _from_date
+    AND   r.valid_to <= _to_date
+    UNION ALL
+    -- Case 1:
+    -- _from_date, _to_date:                  |---------------------------|
+    -- Resource present:                         |-----------------|
+    -- Resource present:                      |---------------------------|
+    SELECT  r.valid_from,
+            r.valid_to,
+            r.resource_guid,
+            r.resource_type,
+            r.resource_name,
+            r.org_guid,
+            r.org_name,
+            r.space_guid,
+            r.space_name,
+            r.plan_name,
+            r.plan_guid,
+            r.storage_in_mb,
+            r.memory_in_mb,
+            r.number_of_nodes
+    FROM  resources r
+    WHERE r.valid_from >= _from_date
+    AND   r.valid_from < _to_date
+    AND   r.valid_to > _from_date
+    AND   r.valid_to <= _to_date
+    UNION ALL
+    -- Case 2:
+    -- _from_date, _to_date:                  |---------------------------|
+    -- Resource present:                                       |-----------------|
+    SELECT  GREATEST(_from_date, r.valid_from),
+            LEAST(_to_date, r.valid_to),
+            r.resource_guid,
+            r.resource_type,
+            r.resource_name,
+            r.org_guid,
+            r.org_name,
+            r.space_guid,
+            r.space_name,
+            r.plan_name,
+            r.plan_guid,
+            r.storage_in_mb,
+            r.memory_in_mb,
+            r.number_of_nodes
+    FROM  resources r
+    WHERE r.valid_from >= _from_date
+    AND   r.valid_from < _to_date
+    AND   r.valid_to > _to_date
+    UNION ALL
+    -- Case 3:
+    -- _from_date, _to_date:                  |---------------------------|
+    -- Resource present:            |---------------------------------------------|
+    SELECT  _from_date,
+            _to_date,
+            r.resource_guid,
+            r.resource_type,
+            r.resource_name,
+            r.org_guid,
+            r.org_name,
+            r.space_guid,
+            r.space_name,
+            r.plan_name,
+            r.plan_guid,
+            r.storage_in_mb,
+            r.memory_in_mb,
+            r.number_of_nodes
+    FROM  resources r
+    WHERE r.valid_from < _from_date
+    AND   r.valid_to > _to_date;
+
+    RETURN QUERY
+    SELECT * FROM calculate_bill();
+END
+$$;
```

Comment:
It's often good style to have each database object, with the exception of simple dependent objects such as indices, constraints, etc., defined in their own file. Not so important with Postgres but can trip you up and cause problems when using other databases